Add Template Engine README and update links in documentation/TemplateEngine folder (#54169)
dotnet/sdk: documentation/TemplateEngine/api/Inside-the-Template-Engine.md

can be time consuming, so consider creating one instance for duration of hosting
application.

## [TemplateCreator](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs) class

The class responsible for template creation, including dry run. The host need to
instantiate the class when needed.

### Supplying parameters conditions results

> It is possible to supply evaluation results of parameters conditions when instantiating template via Edge API [`TemplateCreator.InstantiateAsync`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs#L89). Example use case is instantiation from Visual Studio host, that will leverage condition evaluator integrated within the New Project Dialog. 

Author: ```suggestion
It is possible to supply evaluation results of parameters conditions when instantiating template via Edge API [`TemplateCreator.InstantiateAsync`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs#L76). Example use case is instantiation from Visual Studio host, that will leverage condition evaluator integrated within the New Project Dialog. 
```

> It is possible to supply evaluation results of parameters conditions when instantiating template via Edge API [`TemplateCreator.InstantiateAsync`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs#L76). Example use case is instantiation from Visual Studio host, that will leverage condition evaluator integrated within the New Project Dialog. 

Author: fixed in https://github.com/dotnet/sdk/pull/54169/commits/54c59485d5a48a66a8d01b883ce1ce1155201903

This can be achieved by passing the structured [`InputDataSet`](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/InputDataSet.cs) argument that is populated with [`EvaluatedInputParameterData`](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/EvaluatedInputParameterData.cs) objects for evaluated parameters. 

It is currently not possible to provide just partial external evaluation - meaning that the template engine evaluates either all the parameter conditions or none. If the [`InputDataSet`](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/InputDataSet.cs) collection contains at least one [`EvaluatedInputParameterData`](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/EvaluatedInputParameterData.cs) element, results of all parameter conditions are expected to be passed.

> Template engine cross checks externally passed evaluations. If it encounteres mismatch between externally passed result and internal evaluation result a failed `ITemplateCreationResult` is returned from `InstantiateAsync` API. Failure is indicated by [`CondtionsEvaluationMismatch`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/CreationResultStatus.cs#L61) in [`Status`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/ITemplateCreationResult.cs#L41) property.

Author: ```suggestion
Template engine cross checks externally passed evaluations. If it encounteres mismatch between externally passed result and internal evaluation result a failed `ITemplateCreationResult` is returned from `InstantiateAsync` API. Failure is indicated by [`CondtionsEvaluationMismatch`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/CreationResultStatus.cs#L57) in [`Status`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/ITemplateCreationResult.cs#L39) property.
```

> Template engine cross checks externally passed evaluations. If it encounteres mismatch between externally passed result and internal evaluation result a failed `ITemplateCreationResult` is returned from `InstantiateAsync` API. Failure is indicated by [`CondtionsEvaluationMismatch`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/CreationResultStatus.cs#L57) in [`Status`](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Edge/Template/ITemplateCreationResult.cs#L39) property.

Author: fixed in https://github.com/dotnet/sdk/pull/54169/commits/54c59485d5a48a66a8d01b883ce1ce1155201903

    – mount point implementation, via
    [IMountPointFactory](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/Mount/IMountPointFactory.cs)

-   [IInstaller](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/Installer/IInstaller.cs)
    – template package installer, via
    [IInstallerFactory](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/Installer/IInstallerFactory.cs)

-   [ITemplateConstraint](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/Constraints/ITemplateConstraint.cs)
    – constraint implementation, via
    [ITemplateConstraintFactory](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/Constraints/ITemplateConstraintFactory.cs)

-   [IBindSymbolSource](https://github.com/dotnet/sdk/blob/main/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/Components/IBindSymbolSource.cs)
    – bind symbol source implementation

It is possible to register additional components to in the following way:

-   Components of

>     [ITemplateEngineHost.BuiltInComponents](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs#L21)

Author: ```suggestion
    [ITemplateEngineHost.BuiltInComponents](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs#L17)
```

>     [ITemplateEngineHost.BuiltInComponents](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs#L17)

Author: fixed in https://github.com/dotnet/sdk/pull/54169/commits/54c59485d5a48a66a8d01b883ce1ce1155201903

    will be added when loading environment settings. For
    DefaultTemplateEngineHost they can be passed to constructor.

-   You can use

>     [IComponentManager.AddComponent](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/IComponentManager.cs#L58)

Author: ```suggestion
    [IComponentManager.AddComponent](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/IComponentManager.cs#L53)
```

>     [IComponentManager.AddComponent](https://github.com/dotnet/sdk/blob/0d8f160e2f7/src/TemplateEngine/Microsoft.TemplateEngine.Abstractions/IComponentManager.cs#L53)

Author: fixed in https://github.com/dotnet/sdk/pull/54169/commits/54c59485d5a48a66a8d01b883ce1ce1155201903
